src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +25,122 @@ class DictionarySegment : public BaseSegment {
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    for(ValueID dictionary_index = ValueID{0}; dictionary_index < _dictionary->size(); ++dictionary_index ) {
+      if(_dictionary->at(dictionary_index) >= value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  };
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const {
+      return lower_bound(static_cast<T>(value));
+  };
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    for(ValueID dictionary_index = ValueID{0} ; dictionary_index < _dictionary->size(); ++dictionary_index ) {
+      if(_dictionary->at(dictionary_index) > value) {
+        return dictionary_index;
+      }
+    }
+
+    return INVALID_VALUE_ID;
+  };
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const {
+      return upper_bound(static_cast<T>(value));
+  };
 
-  // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  // return the number of _dictionary (dictionary entries)
+  size_t unique_values_count() const {
+      return _dictionary->size();
+  };
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const {
+      return _attribute_vector->size();
+  };
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+      return 0;
+  };
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
-  std::shared_ptr<BaseAttributeVector> _attribute_vector;
+  std::shared_ptr<std::vector<uint32_t>> _attribute_vector;
+
+  void _build_compressed_dictionary(const std::vector<T>& values) {
+    auto raw_dictionary = std::move(values);
+    std::vector<T> raw_values = raw_dictionary;
+
+    std::sort(raw_dictionary.begin(), raw_dictionary.end());
+    raw_dictionary.erase(std::unique(raw_dictionary.begin(), raw_dictionary.end()), raw_dictionary.end());
+
+    _dictionary = std::make_shared<std::vector<T>>(raw_dictionary);
+
+    _build_attribute_vector(raw_dictionary, raw_values);
+  }
+
+  void _build_attribute_vector(std::vector<T>& raw_dictionary, std::vector<T>& all_values) {
+    std::vector<uint32_t> raw_vector{};
+    raw_vector.reserve(all_values.size());
+
+    for(auto value: all_values) {
+      auto dictionary_iterator = std::find(raw_dictionary.begin(), raw_dictionary.end(), value);
```

Comment:
The `raw_dictionary`is sorted so we should probably use a more appropriate method like `std::lower_bound`. See here https://riptutorial.com/cplusplus/example/7270/using-a-sorted-vector-for-fast-element-lookup

Reply:
Also, should we do another safety check here in case there is no such value encoded in the dictionary, instead of just appending the unvalidated index ...

Reply:
Good points! Will implement that.

Reply:
The efficient lookup is a good point, however, the safety check is unnecessary, because the vector has to contain the value by definition, since it was created from all the values we are iterating over here.

Reply:
safety first sach ich imma :D nee gönnen wa so machen